### PR TITLE
Use IRGen's LinkInfo in TBDGen.

### PR DESCRIPTION
**Pull request by huonw**
Also, use it to generate symbols for conformance witness tables.

### CODE REVIEW

**Comment by huonw** [comment]
@swift-ci Please smoke test.

File: lib/FrontendTool/FrontendTool.cpp [hunk 1]
```diff
@@ -761,10 +762,16 @@ static bool performCompile(std::unique_ptr<CompilerInstance> &Instance,
   }
 
   if (opts.ValidateTBDAgainstIR) {
-    bool validationError =
-        PrimarySourceFile ? validateTBD(PrimarySourceFile, *IRModule)
-                          : validateTBD(Instance->getMainModule(), *IRModule);
-    if (validationError)
+    auto hasMultipleIRGenThreads = Invocation.getSILOptions().NumThreads > 1;
```
Comment (slavapestov) [inline lib/FrontendTool/FrontendTool.cpp]: This setting should not influence the set of public symbols we emit.

File: lib/TBDGen/TBDGen.cpp [hunk 2]
```diff
@@ -83,36 +99,35 @@ class TBDGenVisitor : public ASTVisitor<TBDGenVisitor> {
...
-    auto tmda = irgen::LinkEntity::forTypeMetadataAccessFunction(declaredType);
-    addSymbol(tmda.mangleAsString());
-
-    if (isPrivateDecl(NTD))
-      return;
+        /*isPattern=*/false));
+    addSymbol(irgen::LinkEntity::forTypeMetadataAccessFunction(declaredType));
+
+    // There are symbols associated with any protocols this type conforms to.
+    for (auto conformance : NTD->getAllConformances()) {
```
Comment (slavapestov) [inline lib/TBDGen/TBDGen.cpp]: Isn't there a getLocalConformances() that only returns conformances defined in the same module?

File: lib/TBDGen/TBDGen.cpp [hunk 3]
```diff
@@ -35,23 +36,38 @@ static bool isPrivateDecl(ValueDecl *VD) {
 namespace {
 class TBDGenVisitor : public ASTVisitor<TBDGenVisitor> {
   StringSet &Symbols;
+  const irgen::UniversalLinkageInfo &UniversalLinkInfo;
```
Comment (slavapestov) [inline lib/TBDGen/TBDGen.cpp]: using namespace swift::irgen; would help IMO

File: lib/FrontendTool/FrontendTool.cpp (hunk 1 above)
Comment (huonw) [inline lib/FrontendTool/FrontendTool.cpp]: I'm just passing through all the options that IRGen is asking for ( https://github.com/apple/swift/blob/d1c5de1cea6c388c78266cc1b691074d06b6eb33/lib/IRGen/GenDecl.cpp#L1372-L1386 ), to make sure we're matching as closely as possible. If you think this isn't required I'd be very happy to get rid of it.

**Comment by huonw** [comment]
Updated for the review, except for the multiple threads point. And, also, includes a fix for the PublicExternal thing (first commit).

**Comment by huonw** [comment]
@swift-ci Please smoke test.

**Comment by slavapestov** [comment]
Looks good to me. If we need the multiple threads thing to initialize the IRGen it's fine, is just want to be extra careful that it doesn't change symbol visibility.

**Pull request merged**